Optimize Commit pipeline performance

ethercflow opened the pull request:
Follow up on #267

Optimize the waiting delay caused by the uneven load of the writer wb: the writer who enters the request queue later enters the commit phase first.

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

ethercflow committed:
Optimize Commit pipeline performance

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

ethercflow commented:
@Connor1996 @tabokie  PTAL, thanks!

tabokie commented:
Please add some descriptions for the new algorithm. I didn't find any on the weekly report.

ethercflow commented:
> Please add some descriptions for the new algorithm. I didn't find any on the weekly report.

1. Create a task queue for each writer, the difference with `MultiWriteBatch` is that the latter shares the queue；
2. After each writer confirms the order of submission in the request queue, it loops through its own task queue. The definition of each task is the same as that of MultiWriteBatch；
3. If the head of the request queue completes the memtable write operation first and enters the commit stage, it can return directly without waiting for other writers. This behavior is consistent with the previous pipelined commit；
4. If the non-head writers first enter the commit phase and find that the task queue of the head writer has pending tasks, the loop will take out tasks from the task queue of the head writer for execution；
4.1 After the header is completed, it is found that the helpers are also completed, then pop these writers from the request queue in turn, and notify the helpers to return；
4.2 If the header completes, but finds that the tasks the helpers performed for him have not yet been processed, wait for the helpers to complete and wake it up to clean up from the request queue.

Sorry I missed the email, I've add the descriptions. Please let me know if it's clear enough, thanks! @tabokie @Connor1996

tabokie commented on util/safe_queue.h:
```diff
@@ -0,0 +1,63 @@
+#include <queue>
+
+namespace rocksdb {
+class SafeFuncQueue {
```
I think introducing another queue structure is unnecessary. Instead we can directly iterate the vector of write batches. Something like this:
```c++
class Writer {
 public:
  void SetContext(SequenceNumber sequence,
    ColumnFamilySet* version_set,
    FlushScheduler* flush_scheduler,
    bool ignore_missing_column_families,
    DB* db);
  bool ConsumeOne() {
    claimed_cnt += 1;
    auto b = self.batches[claimed_cnt];
    /* do write */
    consumed_cnt += 1;
  }
 private:
  vector<WriteBatch*> batches;
  std::atomic<uint64_t> claimed_cnt{0};
  std::atomic<uint64_t> consumed_cnt{0};
}
```

ethercflow replied:
Sure,  I'll change to this, thanks!

tabokie commented on db/write_thread.h:
```diff
@@ -459,4 +481,13 @@ class WriteThread {
+struct CommitRequest {
+  WriteThread::Writer* writer;
+  uint64_t commit_lsn;
+  std::atomic<bool> applied;
```
I suspect this field is no longer useful.

ethercflow replied:
The current implementation relies on this, but the variable name is not well expressed in this scenario. 

It is used to prevent when:

Request queue has A and B, A is the first writer, and runs very fast entered commit, but B is just set itself into the request queue. If remove `applied`,  after A pop itself from request queue, it will also commit B directly.

I'll change this when changed to iter wb array.

tabokie replied:
Why? I think `running_nr` can do the job. More specifically, you can set the `running_nr` before adding req to the request queue.

ethercflow replied:
indeed :)

Connor1996 commented on db/db_impl/db_impl.h:
```diff
@@ -154,6 +154,10 @@ class DBImpl : public DB {
                        WriteBatch* updates) override;
 
+  using DB::PebbleWrite;
+  virtual Status PebbleWrite(const WriteOptions& options,
```
maybe use previous multibatchwrite?

ethercflow replied:
Agree, `multibatchwrite` seems better

Connor1996 commented on db/write_thread.cc:
```diff
@@ -500,24 +500,26 @@ void WriteThread::EnterAsMemTableWriter(Writer* leader,
-      if (w->batch->HasMerge()) {
+      if (w->batches[0]->HasMerge()) {
+	#include <cstdio>
+	fprintf(stderr, "here?\n");
```
clean it?

ethercflow replied:
Oh, sorry for this

Connor1996 commented on db/write_thread.cc:
```diff
@@ -781,13 +784,33 @@ void RequestQueue::CommitSequenceAwait(CommitRequest* req,
+        guard.unlock();
+        task();
+        guard.lock();
+	continue;
```
wrong indent

Connor1996 commented:
> 

> > Please add some descriptions for the new algorithm. I didn't find any on the weekly report.
> 
> 1. Create a task 
>            
>              queue
>              (队列)
>            
>             for each writer, the difference with `MultiWriteBatch` is that the latter shares the 
>            
>              queue
>              (队列)
>            
>            ；
> 2. After each writer confirms the order of 
>            
>              submission
>              (提交)
>            
>             in the request 
>            
>              queue
>              (队列)
>            
>            , it loops through its own task 
>            
>              queue
>              (队列)
>            
>            . The definition of each task is the same as that of MultiWriteBatch；
> 3. If the head of the request 
>            
>              queue
>              (队列)
>            
>             completes the memtable write operation first and enters the commit stage, it can return directly without waiting for other writers. This behavior is consistent with the previous pipelined commit；
> 4. If the non-head writers first enter the commit phase and find that the task 
>            
>              queue
>              (队列)
>            
>             of the head writer has pending tasks, the loop will take out tasks from the task 
>            
>              queue
>              (队列)
>            
>             of the head writer for execution；
>    4.1 After the 
>            
>              header
>              (标头)
>            
>             is completed, it is found that the 
>            
>              helpers
>              (帮手)
>            
>             are also completed, then pop these writers from the request 
>            
>              queue
>              (队列)
>            
>             in turn, and notify the 
>            
>              helpers
>              (帮手)
>            
>             to return；
>    4.2 If the 
>            
>              header
>              (标头)
>            
>             completes, but finds that the tasks the 
>            
>              helpers
>              (帮手)
>            
>             performed for him have not yet been processed, wait for the 
>            
>              helpers
>              (帮手)
>            
>             to complete and wake it up to clean up from the request 
>            
>              queue
>              (队列)
>            
>            .
> 
> Sorry I missed the email, I've add the descriptions. Please let me know if it's clear enough, thanks! @tabokie @Connor1996

How about adding these as code comment

ethercflow commented:
> > 
> 
> > > Please add some descriptions for the new algorithm. I didn't find any on the weekly report.
> > 
> > 
> > 
> > 1. Create a task
> >    ```
> >          queue
> >          (队列)
> >        
> >         for each writer, the difference with `MultiWriteBatch` is that the latter shares the 
> >        
> >          queue
> >          (队列)
> >        
> >        ；
> >    ```
> > 2. After each writer confirms the order of
> >    ```
> >          submission
> >          (提交)
> >        
> >         in the request 
> >        
> >          queue
> >          (队列)
> >        
> >        , it loops through its own task 
> >        
> >          queue
> >          (队列)
> >        
> >        . The definition of each task is the same as that of MultiWriteBatch；
> >    ```
> > 3. If the head of the request
> >    ```
> >          queue
> >          (队列)
> >        
> >         completes the memtable write operation first and enters the commit stage, it can return directly without waiting for other writers. This behavior is consistent with the previous pipelined commit；
> >    ```
> > 4. If the non-head writers first enter the commit phase and find that the task
> >    ```
> >          queue
> >          (队列)
> >        
> >         of the head writer has pending tasks, the loop will take out tasks from the task 
> >        
> >          queue
> >          (队列)
> >        
> >         of the head writer for execution；
> >    ```
> >    
> >    
> >        
> >          
> >        
> >    
> >          
> >        
> >    
> >        
> >      
> >    4.1 After the
> >    ```
> >          header
> >          (标头)
> >        
> >         is completed, it is found that the 
> >        
> >          helpers
> >          (帮手)
> >        
> >         are also completed, then pop these writers from the request 
> >        
> >          queue
> >          (队列)
> >        
> >         in turn, and notify the 
> >        
> >          helpers
> >          (帮手)
> >        
> >         to return；
> >    ```
> >    
> >    
> >        
> >          
> >        
> >    
> >          
> >        
> >    
> >        
> >      
> >    4.2 If the
> >    ```
> >          header
> >          (标头)
> >        
> >         completes, but finds that the tasks the 
> >        
> >          helpers
> >          (帮手)
> >        
> >         performed for him have not yet been processed, wait for the 
> >        
> >          helpers
> >          (帮手)
> >        
> >         to complete and wake it up to clean up from the request 
> >        
> >          queue
> >          (队列)
> >        
> >        .
> >    ```
> > 
> > Sorry I missed the email, I've add the descriptions. Please let me know if it's clear enough, thanks! @tabokie @Connor1996
> 
> How about adding these as code comment

Sure, thanks!

Connor1996 commented on util/safe_queue.h:
```diff
@@ -0,0 +1,63 @@
+class SafeFuncQueue {
+ public:
+  std::atomic<size_t> running_nr;
+ private:
```
merge private parts together?

Connor1996 commented on util/safe_queue.h:
```diff
@@ -0,0 +1,63 @@
+      return false;
+    }
+    auto func = std::move(que_.front().func);
+    que_.pop_front();
```
reuse `this->Pop()`?

Connor1996 commented on db/write_thread.h:
```diff
@@ -139,6 +134,7 @@ class WriteThread {
   // Information kept for every waiting writer.
   struct Writer {
     WriteBatch* batch;
```
remove it?

ethercflow committed:
Remove safe queue, iterate the wbs directly

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

ethercflow commented:
Hi @tabokie @Connor1996  I've updated a version according to your suggestions (except for the supplementary comments, which will be added later),  PTAL again, thank you.

tabokie commented on table/table_test.cc:
```diff
@@ -1189,7 +1189,7 @@ TEST_F(TablePropertyTest, PrefixScanTest) {
 
   // prefixes that exist
-  for (const std::string& prefix : {"num.111", "num.333", "num.555"}) {
+  for (const std::string prefix : {"num.111", "num.333", "num.555"}) {
```
revert these changes.

ethercflow replied:
It will fail caused by `-Wrange-loop-analysis`, so I changed it.

tabokie commented on db/db_impl/db_impl_write.cc:
```diff
@@ -80,14 +79,23 @@ void DBImpl::PebbleWriteCommit(CommitRequest* request) {
+  }
+}
+
 Status DBImpl::PebbleWriteImpl(const WriteOptions& write_options,
```
Rename as well.

tabokie commented on db/db_impl/db_impl_write.cc:
```diff
@@ -80,14 +79,23 @@ void DBImpl::PebbleWriteCommit(CommitRequest* request) {
 
+Status DBImpl::MultiBatchWrite(const WriteOptions& options,
+                           std::vector<WriteBatch*>&& updates) {
+  if (immutable_db_options_.enable_pipelined_commit) {
```
Actually, I believe name should be consistent. How about enable_multi_batch_write?

tabokie replied:
And please add some description for the algorithm (and maybe credit to pebble) in the comments for this flag.

ethercflow replied:
Done, thanks!

tabokie commented on db/write_thread.h:
```diff
@@ -136,9 +133,46 @@ class WriteThread {
+      db = _db;
+    }
+
+    void ConsumeOne(Writer* writer, size_t claimed);
+    bool ConsumeOne(Writer* writer);
```
I prefer to see them implemented in header (they are small and very self-contained). Same goes for functions of `RequestQueue`.

tabokie replied:
Also, I find it a little weird that it requires a `Writer*` input when itself is a member of Writer. How about implementing this in `Writer`?

ethercflow replied:
> all and very self-contained). Same goes for functions of `RequestQueue`.


I tried, but met 

```
./db/write_thread.h:171:18: error: ‘WriteBatchInternal’ has not been declared
  171 |       Status s = WriteBatchInternal::InsertInto(
```

As `WriteBatchInternal`  defined in `db/write_batch_internal.h`, which include `db/write_thread.h`, So I think it cannot be implemented in header.

ethercflow replied:
The `RequestQueue` and `WriteThead`  are mutual dependencies, so they cannot be placed in the same header file.

tabokie commented on db/write_thread.cc:
```diff
@@ -781,13 +782,33 @@ void RequestQueue::CommitSequenceAwait(CommitRequest* req,
+    // When the subsequent commit finds that the front writer has not yet
+    // submitted, it will help the front writer to perform some tasks
+    auto front = requests_.front()->writer;
+    if (front->multi_batch_writer.pending_wb_cnt.load(std::memory_order_acquire) > 1) {
+      auto claimed = front->multi_batch_writer.claimed_cnt.fetch_add(1, std::memory_order_acquire);
+      if (claimed < front->multi_batch_writer.batches.size()) {
+        guard.unlock();
+        front->multi_batch_writer.ConsumeOne(front, claimed);
+        guard.lock();
+        continue;
+      }
+    } else {
+      // The front writer may be waiting for this helper writer
+      commit_cv_.notify_all();
+    }
```
If there a reason why this doesn't work?
```suggestion
    guard.unlock();
    bool res = front->multi_batch_writer.ConsumeOne(front);
    guard.lock();
    if (!res) {
      commit_cv_.notify_all();
    } else {
      continue;
    }
```

ethercflow replied:
Yes, I think if we `unlock` before we make sure we've already got a valid `claimed`, it may be danger, eg:

There're two writers in `requests queue`,  the front writer thread (short for A),  the other writer (short for B)

T0: `A` has finished its all wbs will enter `CommitSequenceAwait` , `B` is as same as `A`.
T1: B gets lock, A wait for lock.
T2: B gets A's writer (A's stack value), `unlock` but it's interrupted by some reason, eg: IRQ or OS period resched;
T3: A gets lock, pop itself from `requests queue` , then `PebbleWriteImpl` returns to its caller (so `front` has been reclaimed);
T4: B call ` front->multi_batch_writer.ConsumeOne` , at this point the `front`  is already an invalid value;

tabokie replied:
Okay. I suggest move the atomic operations inside a function, e.g. Claim, Consume.

ethercflow replied:
> Okay. I suggest move the atomic operations inside a function, e.g. Claim, Consume.

Done, thanks!

ethercflow reviewed (commented):
I have a few questions about how to deal with memtable write error:
1. What will happen if write `WAL` successfully, but write `memtable` failed?
2. We loop write `memtable`, according to the current logic, only the latest error is recorded, it this expected?
3. The current helper thread does not call WriteStatusCheck(writer.status); after helping to execute wb, is this ok? (This function is a protected member defined in class db_impl, which is more troublesome to call in write_thread.cc)

tabokie commented:
1. Do nothing and return the error. RocksDB will ignore the error when replaying logs during recovery (if paranoid_checks is not set).
2. I don't think it's a big issue. Memtable don't err unless user writes are seriously malformed.
3. Could you run the check after commit? at which point all tasks should be completed.

ethercflow committed:
Refactored some code
  - Move atomic ops into functions
  - Unified replacement of PebbleWrite with MultiBatchWrite
  - Add a few comments

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

ethercflow commented:
> 1. Do nothing and return the error. RocksDB will ignore the error when replaying logs during recovery (if paranoid_checks is not set).
> 2. I don't think it's a big issue. Memtable don't err unless user writes are seriously malformed.

Got it, thanks!

> 3. Could you run the check after commit? at which point all tasks should be completed.

I think it's OK.

tabokie commented on db/db_impl/db_impl_write.cc:
```diff
@@ -122,8 +162,17 @@ Status DBImpl::PebbleWriteImpl(const WriteOptions& write_options,
+            size_t count = WriteBatchInternal::Count(w->multi_batch_writer.batches);
             if (count > 0) {
+              auto sequence = w->sequence;
+              auto pending_wb_cnt = w->multi_batch_writer.pending_wb_cnt.load();
+              for (size_t i = 0; i < pending_wb_cnt; i++) {
+                auto b = w->multi_batch_writer.batches[i];
```
```suggestion
              for (auto b: w->multi_batch_writer.batches) {
```

tabokie commented on db/version_set.cc:
```diff
@@ -4889,8 +4889,10 @@ Status VersionSet::DumpManifest(Options& options, std::string& dscname,
                      cfd->GetName().c_str(), cfd->GetID());
-              printf("%s at level %d\n", files[i]->DebugString(hex).c_str(),
-                     level);
+              auto debugged_string = files[i]->DebugString(hex);
```
Revert?

ethercflow replied:
Strange, it seems to be generated by I rebase 6.4.tikv, I'll fix it

tabokie commented on db/write_thread.cc:
```diff
@@ -781,13 +782,32 @@ void RequestQueue::CommitSequenceAwait(CommitRequest* req,
+    // When the subsequent commit finds that the front writer has not yet
+    // submitted, it will help the front writer to perform some tasks
+    auto front = requests_.front()->writer;
+    if (front->ConsumableOnOtherThreads()) {
```
If I remember right, it's possible that during memtable write of current group, another write group can be formed to write WAL and added into the request queue. Previously it's not an issue because we do nothing unless the request is committed. But now you might consume those requests whose data isn't written into WAL yet.

ethercflow replied:
> it's possible that during memtable write of current group (A), another write group (B) can be formed to write WAL and added into the request queue. 

So A must have finished wrinting WAL right?

> But now you might consume those requests whose data isn't written into WAL yet.

I think this won't happen, as when A finished its memtable,  A returns directly, will not help B (only the following help the first)  no matter what stage B is in.

I think If there is no WAL allowed to write in parallel, there is no problem. Am I understanding some information wrong?

tabokie replied:
Then you need to check self.req.committed before entering this while loop. Because self.req might already be dequeued and it will loop to write group B.

ethercflow replied:
I can't figure out how it could happen.

tabokie replied:
self.req.committed is already checked. My mistake.

tabokie commented on db/db_impl/db_impl_write.cc:
```diff
@@ -189,23 +238,18 @@ Status DBImpl::PebbleWriteImpl(const WriteOptions& write_options,
-    PebbleWriteCommit(writer.request);
   } else if (writer.request->commit_lsn != 0) {
-    PebbleWriteCommit(writer.request);
+    writer.multi_batch_writer.pending_wb_cnt.store(0);
```
It's possible that a write with !status.ok() can still be written to memtable by another writer.

ethercflow replied:
Yes It's possible.

tabokie replied:
I think this is dangerous. How about let leader reset the failed writer's `pending_wb_cnt`, before waking up write group?

ethercflow replied:
Sure

tabokie reviewed (commented):
I suggest adding another atomic flag `persisted` to mark that a request is safe to be applied. It can solve the two issues in my comments.

ethercflow commented:
> 

Sure, thanks a lot!

ethercflow committed:
Reset pending_wb_cnt before wakeup writers when write WAL failed

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

ethercflow committed:
Update db/db_impl/db_impl_write.cc

Co-authored-by: Xinye Tao <xy.tao@outlook.com>
Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

tabokie commented on db/db_impl/db_impl_write.cc:
```diff
@@ -172,6 +219,8 @@ Status DBImpl::PebbleWriteImpl(const WriteOptions& write_options,
     if (writer.status.ok()) {
       pending_memtable_writes_ += memtable_write_cnt;
+    } else {
+      writer.multi_batch_writer.pending_wb_cnt.store(0);
```
Add some comments to explain we must reset before exit.

tabokie commented on db/db_impl/db_impl_write.cc:
```diff
@@ -189,23 +238,17 @@ Status DBImpl::PebbleWriteImpl(const WriteOptions& write_options,
+    MultiBatchWriteCommit(writer.request);
   } else {
-    writer.request->applied.store(true, std::memory_order_release);
+    writer.multi_batch_writer.pending_wb_cnt.store(0);
```
Put this into a function as well.

tabokie commented on db/write_batch_internal.h:
```diff
@@ -196,6 +196,11 @@ class WriteBatchInternal {
                            bool seq_per_batch = false, size_t batch_cnt = 0,
                            bool batch_per_txn = true);
 
+  static Status InsertInto(
```
Remind me why add this?

ethercflow replied:
As I need to call `InsertInto` with `WriteBatch` , I found  
``` c
Status WriteBatchInternal::InsertInto(
    const WriteBatch* batch, ColumnFamilyMemTables* memtables,
    FlushScheduler* flush_scheduler, bool ignore_missing_column_families,
    uint64_t log_number, DB* db, bool concurrent_memtable_writes,
    SequenceNumber* next_seq, bool* has_valid_writes, bool seq_per_batch,
    bool batch_per_txn);
```
is only used by `repair` and it doesn't call 
`
inserter.set_log_number_ref(writer->log_ref);
`
 before 
`
 Status s = batch->Iterate(&inserter);
`

So I add this one.

ethercflow replied:
As I need to call `InsertInto` with `WriteBatch` , I found  
``` c
Status WriteBatchInternal::InsertInto(
    const WriteBatch* batch, ColumnFamilyMemTables* memtables,
    FlushScheduler* flush_scheduler, bool ignore_missing_column_families,
    uint64_t log_number, DB* db, bool concurrent_memtable_writes,
    SequenceNumber* next_seq, bool* has_valid_writes, bool seq_per_batch,
    bool batch_per_txn);
```
is only used by `repair` and it doesn't call 
`
inserter.set_log_number_ref(writer->log_ref);
`
 before 
`
 Status s = batch->Iterate(&inserter);
`

So I add this one.

tabokie replied:
I think it's only used for transactional DB. How about adding a log_ref parameter to the existing function?

tabokie commented on db/write_thread.cc:
```diff
@@ -797,4 +817,28 @@ void RequestQueue::CommitSequenceAwait(CommitRequest* req,
+  multi_batch_writer.pending_wb_cnt--;
+}
+
+bool WriteThread::Writer::ConsumeOne() {
```
This function can be inlined in header.

tabokie commented on db/write_thread.cc:
```diff
@@ -781,13 +782,32 @@ void RequestQueue::CommitSequenceAwait(CommitRequest* req,
-           requests_.front()->applied.load(std::memory_order_acquire)) {
+    // Because the front writer can return after committing,
+    // so just wait for the help writers to finish
+    while (requests_.front()->writer->HasPendingWB()) {
```
```suggestion
    // As the front writer, some write tasks can be stolen by other writers.
    // Wait for them to finish.
    while (req->HasPendingWB()) {
```

tabokie commented on db/write_thread.cc:
```diff
@@ -797,4 +817,28 @@ void RequestQueue::CommitSequenceAwait(CommitRequest* req,
+    std::lock_guard<std::mutex> guard(this->StateMutex());
+    this->status = s;
+  }
+  multi_batch_writer.pending_wb_cnt--;
```
Pls be specific about memory order.

tabokie commented on db/write_thread.h:
```diff
@@ -271,6 +326,22 @@ class WriteThread {
     }
+
+    bool ConsumableOnOtherThreads() {
+      return multi_batch_writer.pending_wb_cnt.load(std::memory_order_acquire) > 1;
```
Should probably reuse `HasPendingWB`. 1 here doesn't really means anything.

Connor1996 commented on db/write_thread.h:
```diff
@@ -136,9 +133,43 @@ class WriteThread {
     Iterator end() const { return Iterator(nullptr, nullptr); }
   };
 
+  struct MultiBatchWriter {
```
maybe just call it `MultiBatch`?

Connor1996 reviewed (approved):
rest LGTM

ethercflow committed:
Update db/write_thread.cc

Co-authored-by: Xinye Tao <xy.tao@outlook.com>
Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

ethercflow committed:
Refactor some code

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

tabokie commented:
Might need to cherry-pick to 6.29.tikv branch.

ethercflow committed:
Fix two bugs about write memtable

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

ethercflow commented:
/run-test
